File: src/ReshapeTheMatrix.java
public class ReshapeTheMatrix {
    public int[][] matrixReshape(int[][] mat, int r, int c) {
        if (r * c != mat.length * mat[0].length || (mat.length == r && mat[0].length == c)) return mat;

        int[][] result = new int[r][c];
        for (int i = 0, k = 0 ; i < mat.length ; i++) {
            for (int j = 0 ; j < mat[0].length ; j++, k++) {
                result[k / c][k % c] = mat[i][j];
            }
        }
        return result;
    }
}

File: python/reshape_the_matrix.py
from typing import List


class Solution:
    def matrixReshape(self, mat: List[List[int]], r: int, c: int) -> List[List[int]]:
        if r * c != len(mat) * len(mat[0]) or (r == len(mat) and c == len(mat[0])): return mat
        result = [[0] * c for _ in range(r)]
        k = 0
        for i in range(len(mat)):
            for j in range(len(mat[0])):
                result[k // c][k % c] = mat[i][j]
                k += 1
        return result
